Doc update to introduce Tiered Platform Support Model (#8224)
open-telemetry/opentelemetry-collector: docs/platform-support.md

> # Platform Support

Reviewer: IMO we should define what the different platforms are. To that end I think we can just say they are a pair consisting of an operating system family together with a processor architecture family as defined by the Go team through the GOOS and GOARCH tags. We can also comment on this area as to what operating system versions we support (specifically, I would say as a library we support all OS versions officially supported by Go on their supported versions, and as a binary all the ones on the latest version of Go).

cc @rmfitzpatrick on this, please comment if you think the OS versions supported should be different.

Author: Added definition of platform in this section. With respect to a generic support statement for OS and binaries need to discuss further, as it is not clear to me yet how this relates to test strategy and association with support tiers.

Reviewer: > it is not clear to me yet how this relates to test strategy and association with support tiers

My point is just that if we say we support Linux/Windows, it's unclear what versions of that OS family we support and we should say that as part of the definition of each platform.

Author: Added statement about supported OS versions based on Go support.

> # Platform Support

Reviewer: Maybe not on this document/PR, but would also be good to define the process by which a platform can move between tiers (probably needs Collector approver/maintainer sponsorship?)

Author: Can we assume that the assignment of a platform to a specific support tier is fully and only dependent on how the platform is tested as per test strategy? Can we further assume that a platform is elevated to an higher support tier once all required tests for this tier are coded, automatically executed and successfully completed? In this cases, we may tie the description for moving to an higher tier together with the test strategy. In-fact, maybe reasonable to separate out the test strategy into another document.

Reviewer: I think there are concerns that are not strictly related to testing, e.g. the ones I mention on https://github.com/open-telemetry/opentelemetry-collector/pull/8224#discussion_r1293645840 and https://github.com/open-telemetry/opentelemetry-collector/pull/8224#discussion_r1293642032. Maybe we can tie these to testing, but I am unsure about it (e.g. I don't think with our current release schedule a bug on a well-tested but very niche platform would warrant a bugfix release).

The OpenTelemetry Collector will be supported following a tiered platform support model to balance between the aim to support as many platforms as possible and to guarantee stability for the most important platforms. The platform support for the OpenTelemetry Collector is broken into three tiers with different levels of support for each tier  and aligns with the current test strategy.

## Current Test Strategy

> 

Reviewer: I would make this less concrete as to what specific versions we use for testing, outlining instead how we decide these versions. For CI testing this has de facto been "the latest OS version supported on Github runners".

Author: Removed references to specific OS versions.

> The current verification process of the OpenTelemetry Collector includes unit and performance tests for core and additional end-to-end and integration tests for contrib. In the end-to-end tests, receivers, processors, and exporters etc. are tested in a testbed, while the integration tests rely on actual instances and available container images. Additional stability tests are in preparation for the future as well. All verification tests are run on Ubuntu 22.04 Linux on amd64 as the primary platform today. In addition, unit tests are run for the _contrib_ collector on Microsoft Windows Server 2022 (amd64). The cross compile supports two MacOS/Darwin targets (amd64 and arm64), five Linux platforms (amd64, arm64, i386, arm and ppc64le) and two Windows binaries (amd64, i386) today. None of those platforms is tested today, except of Linux on amd64 as the primary platform. The OpenTelemetry Collector can be installed using apk, deb or rpm files and is available as container images (core, contrib) for deployment on Kubernetes and Docker. The container images are built using qemu and published to Docker Hub and ghcr.io for Linux on amd64, arm64, i386, arm/v7 and ppc64le. The end-to-end test for the _contrib_ container images is run on Ubuntu 22.04 Linux for the Kubernetes versions v1.23 to v1.26.

Reviewer: I would leave packaging and installation methods out of this part. They are not strictly related to testing and we recently added wording here https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions#criteria-for-supported-distributions as to what artifacts we promise to release.

Author: Clean up done of the test strategy section.

> ## Tiered platform support model

Reviewer: We should say what guarantees we give to platforms that we give no support to officially. I think https://github.com/open-telemetry/opentelemetry-collector/pull/6924#pullrequestreview-1250396643 could be a good start.

Author: Added statement on unsupported platforms.

The OpenTelemetry Collector will be supported following a tiered platform support model to balance between the aim to support as many platforms as possible and to guarantee stability for the most important platforms. The platform support for the OpenTelemetry Collector is broken into three tiers with different levels of support for each tier. 

### Tier 1 – Primary Support

> The Tier 1 supported platforms are _guaranteed to work_. Precompiled binaries are built on the platform, fully supported for all collector add-ons (receivers, processor, exporters etc.), and continuously tested as part of the development processes to ensure any proposed change will function correctly. Build and test infrastructure is provided by the project. All tests are executed on the platform as part of automated continuous integration (CI) for each pull request and the biweekly release cycle. Any build or test failure block the release of the collector distribution for all platforms. Defects are addressed with priority and depending on severity fixed for the previous release in a bug fix release.

Reviewer: We should state whether a defect specific to a platform of each tier can block the release process or warrant a bugfix release (I think for tier 1 they can).

> The Tier 1 supported platforms are _guaranteed to work_. Precompiled binaries are built on the platform, fully supported for all collector add-ons (receivers, processor, exporters etc.), and continuously tested as part of the development processes to ensure any proposed change will function correctly. Build and test infrastructure is provided by the project. All tests are executed on the platform as part of automated continuous integration (CI) for each pull request and the biweekly release cycle. Any build or test failure block the release of the collector distribution for all platforms. Defects are addressed with priority and depending on severity fixed for the previous release in a bug fix release.

Reviewer: I think this would benefit from a table form presentation, where we have columns for each aspect (CI testing, running tests, bugfix releases, binaries availability, possibility of dummy implementations....) and rows for each tier.

Author: Added summary table.

> The Tier 1 supported platforms are _guaranteed to work_. Precompiled binaries are built on the platform, fully supported for all collector add-ons (receivers, processor, exporters etc.), and continuously tested as part of the development processes to ensure any proposed change will function correctly. Build and test infrastructure is provided by the project. All tests are executed on the platform as part of automated continuous integration (CI) for each pull request and the biweekly release cycle. Any build or test failure block the release of the collector distribution for all platforms. Defects are addressed with priority and depending on severity fixed for the previous release in a bug fix release.

Reviewer: Another requirement for tier 1 platforms in my mind would be the need to have a non-dummy implementation for any feature that is not platform specific. That is, we will not accept PRs that do not implement functionality on e.g. Linux/amd64.

Tier 1 platforms are currently:

> - Linux amd64

Reviewer: I think we should state an owner for each platform supported. For Tier 1 platforms this can be the OTel collector approvers (at least for now), and we can leave the rest empty and create an issue to add owners based on who contributed PRs related to it.

Author: Added table with owners column to supported platforms.

> - Kubernetes amd64

Reviewer: I think our support for Kubernetes comes from specific components that provide functionality related to it, but there is nothing special on the core code to support this. In contrast, to support an operating system family there are changes as to what system APIs are used. So I would be leaning towards not including Kubernetes within this document, or at least separating it somehow.

Author: Removed references to Kubernetes and Docker.

### Tier 2 – Secondary Support

> Tier 2 platforms are _guaranteed to work with specified limitations_. Precompiled binaries are built and tested on the platform as part of the biweekly release cycle. Build and test infrastructure is provided by the platform maintainers. All tests are executed on the platform as far as they are applicable, and all prerequisites are fulfilled. Not executed tests and not tested collector add-ons (received, processors, exporters, etc.) are published on release of the collector distribution. Any build or test failure delays the release of the binaries for the respective platform but not the collector distribution for all other platforms. Defects are addressed but not with the priority as for Tier 1 and, if specific to the platform, require the support of the platform maintainers.

Reviewer 2: ```suggestion
Tier 2 platforms are _guaranteed to work with specified limitations_. Precompiled binaries are built and tested on the platform as part of the biweekly release cycle. Build and test infrastructure is provided by the platform maintainers. All tests are executed on the platform as far as they are applicable, and all prerequisites are fulfilled. Not executed tests and not tested collector add-ons (receivers, processors, exporters, etc.) are published on release of the collector distribution. Any build or test failure delays the release of the binaries for the respective platform but not the collector distribution for all other platforms. Defects are addressed but not with the priority as for Tier 1 and, if specific to the platform, require the support of the platform maintainers.
```

Tier 2 platforms are currently:
- None

### Tier 3 - Community Support

Tier 3 platforms are _guaranteed to build_. Precompiled binaries are made available as part of the release process and as result of a cross compile build on Linux amd64 but the binaries are not tested at all. Any build failure delays the release of the binaries for the respective platform but not the collector distribution for all other platforms. Defects are addressed based on community contributions. Core developers might provide guidance or code reviews, but direct fixes may be limited.

Tier 3 platforms are currently:

> - MacOS/Darwin amd64 

Reviewer: I would use `darwin` here and generally stick to the naming given by the Go team.

Author: Updated platform naming.

- MacOS/Darwin arm64
- Linux arm64 
- Linux i386
- Linux arm/7 
- Linux ppc64le 

> - Windows amd64

Reviewer: I would expect Windows/amd64 to be at least Tier 2, why do you think it is tier 1?

Author: Moved windows/amd64 to Tier 2 and clarified Tier 2 requirements.

- Windows i386
- Kubernetes arm64, i386, arm/v7 and ppc64le
- Docker amd64, arm64, i386, arm/v7 and ppc64le

> The proposed additional platforms Linux on s390x (#378) and AIX on ppc64 ([#19195](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19195#issuecomment-1458560971)) will be included into Tier 3 once they're added to the OpenTelemetry Collector as platforms. 

Reviewer 2: ```suggestion
The proposed additional platforms Linux on s390x ([#378](https://github.com/open-telemetry/opentelemetry-collector-releases/issues/378#issuecomment-1662371014)) and AIX on ppc64 ([#19195](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19195#issuecomment-1458560971)) will be included into Tier 3 once they're added to the OpenTelemetry Collector as platforms. 
```

Author: Instead of linking to specific issue comments here, it is sufficient to link to the related issues. I will update the link for AIX on ppc64.

# Platform Support

> The OpenTelemetry Collector will be supported following a tiered platform support model to balance between the aim to support as many platforms as possible and to guarantee stability for the most important platforms. A platform is described by the pair of operating system and processor architecture family as they are defined by the Go programming language as [known operating systems and architectures for use with the GOOS and GOARCH values](https://go.dev/src/go/build/syslist.go). For a supported platform, the OpenTelemetry Collector is supported when the [minimum requirements](https://github.com/golang/go/wiki/MinimumRequirements) of the Go release used by the collector are met for the operating system and architecture. Each supported platform requires the naming of designated owners. The platform support for the OpenTelemetry Collector is broken into three tiers with different levels of support for each tier and aligns with the current test strategy. For platforms not listed as supported by any of the tiers, support cannot be assumed to be provided. While the project may accept specific changes related to these platforms, there will be no official builds, support of issues and development of enhancements or bug fixes for these platforms. Future development of project for supported platforms may break the functionality of unsupported platforms.

Reviewer: nit: I think we can split this into three paragraphs and make it a bit more readable

```suggestion
The OpenTelemetry Collector will be supported following a tiered platform support model to balance between the aim to support as many platforms as possible and to guarantee stability for the most important platforms. A platform is described by the pair of operating system and processor architecture family as they are defined by the Go programming language as [known operating systems and architectures for use with the GOOS and GOARCH values](https://go.dev/src/go/build/syslist.go). 

For a supported platform, the OpenTelemetry Collector is supported when the [minimum requirements](https://github.com/golang/go/wiki/MinimumRequirements) of the Go release used by the collector are met for the operating system and architecture. Each supported platform requires the naming of designated owners. The platform support for the OpenTelemetry Collector is broken into three tiers with different levels of support for each tier and aligns with the current test strategy. 

For platforms not listed as supported by any of the tiers, support cannot be assumed to be provided. While the project may accept specific changes related to these platforms, there will be no official builds, support of issues and development of enhancements or bug fixes for these platforms. Future development of project for supported platforms may break the functionality of unsupported platforms.
```

## Current Test Strategy

The current verification process of the OpenTelemetry Collector includes unit and performance tests for core and additional end-to-end and integration tests for contrib. In the end-to-end tests, receivers, processors, and exporters etc. are tested in a testbed, while the integration tests rely on actual instances and available container images. Additional stability tests are in preparation for the future as well. All verification tests are run on the linux/amd64 as the primary platform today. In addition, unit tests are run for the _contrib_ collector on windows/amd64. The tests use as execution environments the latest Ubuntu and Windows Server versions [supported as Github runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources). 

The cross compile supports the following targets:
- darwin/amd64 and darwin/arm64
- linux/amd64, linux/arm64, linux/386, linux/arm and linux/ppc64le
- windows/amd64, windows/386. 

Except of the mentioned tests for linux/amd64 and windows/amd64, none of the other platforms are tested today. 

> Container images of the _core_ and _contrib_ collector are built and published to Docker Hub and ghcr.io for linux/amd64, linux/arm64, linux/i386, linux/arm and linux/ppc64le. End-to-end tests of the _contrib_ container images are run on the latest Ubuntu Linux supported by GitHub runners and for the four most recent Kubernetes versions.

Reviewer 3: The list of images may change, as it already has now that it includes s390. I worry this would get out of sync quickly. Maybe it's better to include a link to the goreleaser configuration https://github.com/open-telemetry/opentelemetry-collector-releases/blob/bf8002ec6d2109cdb4184fc6eb6f8bda59ea96a2/.goreleaser.yaml#L137?

Author: Removed the reference to specific platforms for the container images and added reference to the goreleaser configuration.

## Tiered platform support model

> The OpenTelemetry Collector will be supported following a tiered platform support model to balance between the aim to support as many platforms as possible and to guarantee stability for the most important platforms. The platform support for the OpenTelemetry Collector is broken into three tiers with different levels of support for each tier. 

Reviewer 3: ```suggestion
The platform support for the OpenTelemetry Collector is broken into three tiers with different levels of support for each tier. 
```

### Platform Support - Summary

The following tables summarized the platform tiers of support by the verification tests performed for them and by the specification if dummy implementations are allowed for selected features, the availability of precompiled binaries incl. container images and if bugfix releases are provided for previous releases in case of critical defects. 

| Tier | Unit tests | Performance tests | End-to-end tests | Integrations tests | Dummy implementations | Precompiled binaries | Bugfix releases |
|------|------------|-------------------|------------------|--------------------|-----------------------|----------------------|-----------------|
| 1    | yes        | yes               | yes              | yes                | no                    | yes                  | yes             |
| 2    | yes        | optional          | optional         | optional           | yes                   | yes                  | no              |
| 3    | no         | no                | no               | no                 | yes                   | yes                  | no              |

### Tier 1 – Primary Support

> The Tier 1 supported platforms are _guaranteed to work_. Precompiled binaries are built on the platform, fully supported for all collector add-ons (receivers, processor, exporters etc.), and continuously tested as part of the development processes to ensure any proposed change will function correctly. Build and test infrastructure is provided by the project. All tests are executed on the platform as part of automated continuous integration (CI) for each pull request and the biweekly release cycle. Any build or test failure block the release of the collector distribution for all platforms. Defects are addressed with priority and depending on severity fixed for previous release(s) in a bug fix release.

Reviewer 3: ```suggestion
The Tier 1 supported platforms are _guaranteed to work_. Precompiled binaries are built on the platform, fully supported for all collector add-ons (receivers, processor, exporters etc.), and continuously tested as part of the development processes to ensure any proposed change will function correctly. Build and test infrastructure is provided by the project. All tests are executed on the platform as part of automated continuous integration (CI) for each pull request and the [release cycle](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/release.md#release-schedule). Any build or test failure block the release of the collector distribution for all platforms. Defects are addressed with priority and depending on severity fixed for previous release(s) in a bug fix release.
```

Tier 1 platforms are currently:
| Platform    | Owner(s)                                                                                                    |
|-------------|-------------------------------------------------------------------------------------------------------------|
| linux/amd64 | [OpenTelemetry Collector approvers](https://github.com/open-telemetry/opentelemetry-collector#contributing) |

### Tier 2 – Secondary Support

> Tier 2 platforms are _guaranteed to work with specified limitations_. Precompiled binaries are built and tested on the platform as part of the biweekly release cycle. Build and test infrastructure is provided by the platform maintainers. All tests are executed on the platform as far as they are applicable, and all prerequisites are fulfilled. Not executed tests and not tested collector add-ons (receivers, processors, exporters, etc.) are published on release of the collector distribution. Any build or test failure delays the release of the binaries for the respective platform but not the collector distribution for all other platforms. Defects are addressed but not with the priority as for Tier 1 and, if specific to the platform, require the support of the platform maintainers.

Reviewer 3: ```suggestion
Tier 2 platforms are _guaranteed to work with specified limitations_. Precompiled binaries are built and tested on the platform as part of the release cycle. Build and test infrastructure is provided by the platform maintainers. All tests are executed on the platform as far as they are applicable, and all prerequisites are fulfilled. Not executed tests and not tested collector add-ons (receivers, processors, exporters, etc.) are published on release of the collector distribution. Any build or test failure delays the release of the binaries for the respective platform but not the collector distribution for all other platforms. Defects are addressed but not with the priority as for Tier 1 and, if specific to the platform, require the support of the platform maintainers.
```

Tier 3 platforms are currently:
| Platform      | Owner(s)                                                                                                    |
|---------------|-------------------------------------------------------------------------------------------------------------|
| darwin/amd64  |                                                                                                             |
| darwin/arm64  |                                                                                                             |
| linux/arm64   |                                                                                                             |
| linux/386     |                                                                                                             |
| linux/arm     |                                                                                                             |

> | linux/ppc64le |                                                                                                             |

Reviewer 3: ```suggestion
| linux/ppc64le |                                                                                                             |
| linux/s390x   |                                                                                                             |
```

| windows/386   |                                                                                                             |

> The proposed additional platforms linux/s390x (#378) and aix/ppc64 ([#19195](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19195)) will be included into Tier 3 once they're added to the OpenTelemetry Collector as platforms. 

Reviewer 3: ```suggestion
The proposed additional platform aix/ppc64 ([#19195](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19195)) will be included into Tier 3 once it's added to the OpenTelemetry Collector as platform. 
```
